node/src/exchange.rs

```diff
@@ -83,15 +173,356 @@ impl Codec for HeaderCodec {
         &mut self,
         _: &Self::Protocol,
         io: &mut T,
-        resp: Self::Response,
+        resps: Self::Response,
     ) -> io::Result<()>
     where
         T: AsyncWrite + Unpin + Send,
     {
-        let data = resp.encode_length_delimited_to_vec();
+        for resp in resps {
+            let data = resp.encode_length_delimited_to_vec();
 
-        io.write_all(data.as_ref()).await?;
+            io.write_all(&data).await?;
+        }
 
         Ok(())
     }
 }
+
+#[derive(Debug, thiserror::Error, PartialEq)]
+pub enum ReadHeaderError {
+    #[error("stream closed while trying to get header length")]
+    StreamClosed,
+    #[error("varint overflow")]
+    VarintOverflow,
+    #[error("request too large: {0}")]
+    ResponseTooLarge(usize),
+}
```

Comment:
Move this after `pub struct HeaderCodec;`

Reply:
moved :+1: